feat: Add setting to enable autoplay

FabianLars opened the pull request:
3 Open tasks/questions i need help with:
- I have no idea how to make the android one configurable
- The macos implementation didn't work the last time we tested the autoplay-0.19 branch. I can't test it myself because either the webview or my whole vm crashses every time media is involved.
- I'm not sure what the default should be. For actual apps enabling autoplay makes sense, but for displaying external URLs not so much, i went with enabling it by default since this is what was used in android already.

Tauri request: https://github.com/tauri-apps/tauri/issues/3478

### What kind of change does this PR introduce?
<!-- Check at least one. If you are introducing a new binding, you must reference an issue where this binding has been proposed, discussed and approved by the maintainers. -->

- [ ] Bugfix
- [x] Feature
- [ ] Docs
- [ ] Code style update
- [ ] Refactor
- [ ] Build-related changes
- [ ] Other, please describe:

### Does this PR introduce a breaking change?
<!-- If yes, please describe the impact and migration path for existing applications in an attached issue. -->

- [ ] Yes
- [x] No

### Checklist
- [ ] This PR will resolve #___
- [x] A change file is added if any packages will require a version bump due to this PR per [the instructions in the readme](https://github.com/tauri-apps/wry/blob/dev/.changes/readme.md).
- [x] I have added a convincing reason for adding this feature, if necessary
- [ ] It can be built on all targets and pass CI/CD. -> Not yet

### Other information

FabianLars committed:
feat: Add setting to enable autoplay.

amrbashir reviewed (commented):
For Android, you need to add a method like `loadUrlMainThread` and generate a rust binding for it and invoke it whenever you need to change this setting

FabianLars commented:
> whenever you need to change this setting

I'd match it with the desktop platforms where this can only be set when creating the webview. I _hoped_ one can modify the class constructor or something but i guess this will work too, thanks 👍 

i'm gonna see if i can implement it at some point, i currently don't have enough disk space to set up the android dev env but i guess i have enough time until someone fixes macos anyway :D

wusyong commented:
@FabianLars It's okay to leave Android unsupported and add document to mention this. I can test others if you want to promote to PR.

FabianLars committed:
android specific notes and browser_args warning

FabianLars committed:
revert macos

FabianLars commented:
Well, the old autoplay-0.19 was once used in semi-production ready apps so windows and linux should work, so that part would be ready for review.

I just added a note about the android specific behavior so that we could potentially fix that in a follow-up PR.

That leaves us with the macOS part which i can't implement/test because my VM can't handle media playback. I just reverted the macos change to something a bit more sane so if someone could try that real quick? I _think_ i had someone test that variant already but worth to double check.
If it doesn't work, i can remove it from this PR like the android implementation if you want 🤷

wusyong commented on src/webview/wkwebview/mod.rs:
```diff
@@ -315,6 +315,12 @@ impl InnerWebView {
+      if attributes.autoplay {
+        let none: NSUInteger = 0;
+        // FIXME: No idea how to make this not crash the webview, we need to set it to `WKAudiovisualMediaTypeNone` which _should_ be `0` https://developer.apple.com/documentation/webkit/wkaudiovisualmediatypes/wkaudiovisualmediatypenone
+        let _: id = msg_send![config, setValue:none forKey:NSString::new("mediaTypesRequiringUserActionForPlayback")];
```
```suggestion
        let _: id = msg_send![config, setMediaTypesRequiringUserActionForPlayback:0];
```

wusyong commented on src/webview/mod.rs:
```diff
@@ -643,6 +657,7 @@ pub trait WebViewBuilderExtWindows {
   ///
   /// By default wry passes `--disable-features=msWebOOUI,msPdfOOUI,msSmartScreenProtection`
   /// so if you use this method, you also need to disable these components by yourself if you want.
+  /// Furthermore you need to add `--autoplay-policy=no-user-gesture-required` if autoplay is enabled.
```
Does this still apply? I saw there's a if statement in webview2's code.

FabianLars replied:
Yes, the additional_browser_args settings will overwrite the autoplay flag similar to how it overwrites wry's default --disable-features stuff.

But i think i messed up the wording a bit, i'll try to make it more clear.

wusyong commented on src/webview/mod.rs:
```diff
@@ -227,6 +227,13 @@ pub struct WebViewAttributes {
+  ///
+  /// ## Platform-specific:
+  ///
+  /// - **Android:** Autoplay will always be enabled on window creation.
```
Maybe we can say it's not supported yet and point to the line to get started.

FabianLars committed:
apply code review

FabianLars committed:
wording

wusyong reviewed (dismissed):
I'm okay with Android not supported yet, so this looks good to me.

amrbashir commented:
I will implement Android

amrbashir committed:
implement android

wusyong commented:
@FabianLars Can you rebase/merge the branch? It's good to go after this.

FabianLars committed:
Merge branch 'dev' into autoplay
